snuba/datasets/configuration/entity_builder.py

```diff
@@ -0,0 +1,130 @@
+from snuba.datasets.configuration.json_schema import V1_ENTITY_SCHEMA
+from snuba.datasets.configuration.loader import load_configuration_data
+from snuba.datasets.configuration.utils import parse_columns
+from snuba.datasets.entity import Entity
+from snuba.datasets.pluggable_entity import PluggableEntity
+from snuba.datasets.storages import StorageKey
+from snuba.datasets.storages.factory import get_storage, get_writable_storage
+from snuba.query.processors import QueryProcessor
+from snuba.query.validation.validators import (
+    EntityRequiredColumnValidator,
+    QueryValidator,
+)
+
+__MAPPINGS_HAVE_BEEN_INITIALIZED = False
+
+_QP_MAPPING: dict[str, Type[QueryProcessor]] = {}
+
+_FUNCTION_MAPPER_MAPPING: Mapping[str, Type[FunctionCallMapper]] = {
+    "simple_func": FunctionNameMapper
+}
+
+_SUB_MAPPER_MAPPING: Mapping[str, Type[SubscriptableReferenceMapper]] = {
+    "subscriptable": SubscriptableMapper
+}
+
+_VALIDATOR_MAPPING: Mapping[str, Type[QueryValidator]] = {
+    "entity_required": EntityRequiredColumnValidator
+}
+
+logger = logging.getLogger("snuba.entity_builder")
+
+
+def _initialize_mappings() -> None:
+    # circular imports 🤪
+    global __MAPPINGS_HAVE_BEEN_INITIALIZED
+    from snuba.datasets.entities.metrics import TagsTypeTransformer
+    from snuba.query.processors.granularity_processor import MappedGranularityProcessor
+    from snuba.query.processors.object_id_rate_limiter import (
+        OrganizationRateLimiterProcessor,
+        ProjectRateLimiterProcessor,
+        ProjectReferrerRateLimiter,
+        ReferrerRateLimiterProcessor,
+    )
+    from snuba.query.processors.quota_processor import ResourceQuotaProcessor
+    from snuba.query.processors.timeseries_processor import TimeSeriesProcessor
+
+    _QP_MAPPING.update(
+        {
+            "transform_tag_types": TagsTypeTransformer,
+            "handle_mapped_granularities": MappedGranularityProcessor,
+            "translate_time_series": TimeSeriesProcessor,
+            "referrer_rate_limit": ReferrerRateLimiterProcessor,
+            "org_rate_limiter": OrganizationRateLimiterProcessor,
+            "project_referrer_rate_limiter": ProjectReferrerRateLimiter,
+            "project_rate_limiter": ProjectRateLimiterProcessor,
+            "resource_quota_limiter": ResourceQuotaProcessor,
+        }
+    )
+
+    __MAPPINGS_HAVE_BEEN_INITIALIZED = True
+
+
+def _build_entity_validators(
+    config_validators: list[dict[str, Any]]
+) -> Sequence[QueryValidator]:
+    return [
+        _VALIDATOR_MAPPING[qv_config["validator"]](**qv_config["args"])  # type: ignore
+        for qv_config in config_validators
+    ]
+
+
+def _build_entity_query_processors(
+    config_query_processors: list[dict[str, Any]],
+) -> Sequence[QueryProcessor]:
+    return [_QP_MAPPING[config_qp["processor"]](**(config_qp["args"] if config_qp.get("args") else {})) for config_qp in config_query_processors]  # type: ignore
```

Comment:
One thing that should also be in next steps is better error handling I think. This will return a generic `KeyError` if someone misconfigures things.

Reply:
yeah, I agree completely